### ISSUE TITLE
Speed up code with domain decompostion and MPI

### ISSUE THREAD

**Issue by Sierd** [openearth/aeolis-python#115]
When domains get larger (more gridpoints) the calculation slows down. Solving the sparse system of equations (Ax=b) accounts for a large portion of this calculation time. This is done here:
https://github.com/openearth/aeolis-python/blob/6fb285e872dca9ad724805ea79b6adef0c7960df/aeolis/model.py#L2201

Common software packages like Delft3D and XBeach use domain decomposition to speed up code by making it parallel in combination with MPI. See for instance page 606 in the Delft3D manual for an explaination:
https://content.oss.deltares.nl/delft3d4/Delft3D-FLOW_User_Manual.pdf

I wonder if we can adopt this strategy as well.

I have included a typical A matrix and B vector to benchmark a potential solution. 
[Ax=b.zip](https://github.com/openearth/aeolis-python/files/11273853/Ax.b.zip)

Chat GTP has proposed some approach :).

![image](https://user-images.githubusercontent.com/14054272/233099081-480b0f40-6cb8-4605-8055-eef9ef824ef4.png)

**Comment by niketagrawal** [comment]
The following possibilities for speeding up the code have been identified:

- Domain decomposition: Partitioning the domain into two marginally overlapping domains and running them as two parallel simulations. This approach may not work well with different types of domains.

- (To be investigated) The sequential solving of the grain fractions in the solvers can be done in parallel. This corresponds to the big for-loop in the solver code.

- (To be investigated) Certain functions in the hydro module could be run in parallel.

**Comment by Sierd** [comment]
I made a a small code to play with which performs the domain decomposition. It does not do parallelization yet. 

@niketagrawal and @manuGil , could you give your opinion and see how this can be made parallel. 

[parallell_solve.zip](https://github.com/openearth/aeolis-python/files/11484498/parallell_solve.zip)

**Comment by manuGil** [comment]
Hello Sierd, I quickly looked into your code. I will provide you with an example tomorrow.

**Comment by manuGil** [comment]
@Sierd 

In Python we can use the `multiprocessing` packages to do domain decomposition. The basic idea is to *prepare* the computation we want to run in parallel, and then generate **n** number of process that will execute each computation separately. 

This is a basic example using a **Pool** of processes to execute the computation. 
### Notice a few things:
1. The example only runs the computation of the sparse matrix in parallel
2. The split of the data still is done manually, and it is a condition for running (1)
3. Data is combined after the parallel computation in the same way that your code. We must wait until all parallel computation are completed to combine the results. 
4. A rule of thumb for the number of processes is  `number of CPU/cores + 1`.
5. Starting each process in the pool takes time. Therefore a parallel computation in this way may only make sense if: 
   - a) the size of input matrix is large enough, or/and 
   - b) the matrix computation is generally faster for smaller matrices, or/and 
   - c) there are enough CPU/cores available for the computations, and 
   - d) the time of dividing data + spawning processes + combining data tends to be less than running a single process for a matrix X.

```python
import scipy.sparse
import numpy as np

# the function that we want to run in parallel
def solve_matrix(A,Ct_i):
    # solve the matrix
    result  = scipy.sparse.linalg.spsolve(A,Ct_i)

    return result

if __name__ == '__main__':
    from multiprocessing import Pool, Value, Array
    import time

    # Load data
    A = scipy.sparse.load_npz('parallell_solve/A.npz')
    yCt_i = np.load('parallell_solve/yCt_i.npy')

    start_time = time.time()

    #  divide matrices in 2 domains
    decs =    [ 
        (A[0:24080,0:24080], yCt_i.flatten()[0:24080]), # first dec
        (A[21371:,21371:], yCt_i.flatten()[21371:]), # second dec
    ]

   # create a pool to handle computations
    with Pool(processes=5) as pool:
        results = pool.starmap(solve_matrix, decs)

   # some required post-processing
    reshaped_results = [result.reshape(80,301) for result in results]
  
    # combine the results
    comb_pool2 = np.vstack((
        reshaped_results[0][0:75,:],
        reshaped_results[1][4:,:]
        ))
```

**Comment by manuGil** [comment]
BTW, there are other strategies in which the computation can be make parallel. The performance for the code above (2 data parts) is not significantly different than with your initial approach. 

```python
spsolve gives 0.71 seconds, this is the benchmark
decompostion in 2 domains solved sequentially gives 0.59 seconds with an accuracy of 2.86e-09 
decompostion in 4 domains solved sequentially gives 0.51 seconds with an accuracy of 9.90e-05 
decompostion using Pool and 2 domains gives 0.05541229248046875 seconds with an accuracy of 2.8622819543125053e-09
```
```python

**Comment by manuGil** [comment]
In the case of splitting input matrices into 4 parts, the time is about the same as for 2 parts. This is because I am testing this in a 4-CPU machine.
```python
decompostion using Pool and 2 domains gives 0.05121183395385742 seconds with an accuracy of 2.8622819543125053e-09
decompostion using Pool and 4 domains gives 0.04767775535583496 seconds with an accuracy of 9.895653754833297e-05
```

**Comment by manuGil** [comment]
# Timing data decomposition
This is an example of splitting matrices into 2 and 4 parts before running the code in parallel using a processing pool. A plot is generated to compare the computation time against a number of iterations.

```python

def solve_matrix(A,Ct_i):
    # solve the matrix
    result  = scipy.sparse.linalg.spsolve(A,Ct_i)

    return result.reshape(80,301)

# similar to the above, but without reshaping,
# only used for the 4 domains
def solve_matrix4(A,Ct_i):
    # solve the matrix
    result  = scipy.sparse.linalg.spsolve(A,Ct_i)

    return result

if __name__ == '__main__':
    from multiprocessing import Pool
    import matplotlib.pyplot as plt

    ######################################
    ## 2 domains
    ######################################

    # Load data
    A = scipy.sparse.load_npz('parallell_solve/A.npz')
    yCt_i = np.load('parallell_solve/yCt_i.npy')

    # a function for the parallelization
    # this is only for compatibility with timeit
    def pool_2():

        # devide matrices in 2 domains
        decs =    [ 
            (A[0:24080,0:24080], yCt_i.flatten()[0:24080]), # first dec
            (A[21371:,21371:], yCt_i.flatten()[21371:]), # second dec
        ]

        with Pool(processes=5) as pool:
            results = pool.starmap(solve_matrix, decs)
    
        # combine the results
        comb_pool2 = np.vstack((
            results[0][0:75,:],
            results[1][4:,:]
            ))
        return comb_pool2
    

    solve_sp = scipy.sparse.linalg.spsolve(A, yCt_i.flatten()).reshape(151,301)

    comb_pool2 = pool_2()
    diff_comb_pool2 = (comb_pool2-solve_sp).max()


    ######################################
    ## 4 domains
    ######################################                                                                                                  

     # Load data
    A = scipy.sparse.load_npz('parallell_solve/A.npz')
    yCt_i = np.load('parallell_solve/yCt_i.npy')

    # a function for the parallelization
    # this is only for compatibility with timeit
    def pool_4():
        # devide matrices in 4 domains
        decs =    [ 
            (A[0:13545,0:13545], yCt_i.flatten()[0:13545]), # first dec
            (A[12040:25585,12040:25585], yCt_i.flatten()[12040:25585]), # second dec
            (A[24080:37625,24080:37625], yCt_i.flatten()[24080:37625]), # third dec
            (A[36120:45451,36120:45451], yCt_i.flatten()[36120:45451]) # fourth dec
        ]
        
        with Pool(processes=5) as pool:
            results = pool.starmap(solve_matrix4, decs)
    
        reshaped_results = [result.reshape(45,301) for result in results[:-1]]
        last_result = results[-1].reshape(31,301)

        # combine the results
        comb_pool4 = np.vstack((
            reshaped_results[0][0:41,:],reshaped_results[1][1:-5,:],reshaped_results[2][:-5,:],last_result[:,:]
            ))

        return comb_pool4

    comb_pool4 = pool_4()
    
    diff_comb_pool4 = (comb_pool4-solve_sp).max()

    print('decompostion using Pool with 2 domains solved sequentially gives %.2f seconds with an accuracy of %.2e ' %  ( timeit.timeit('comb_pool2 = pool_2()', number=1,globals=globals()), diff_comb_pool2 ) )
    print('decompostion using Pool with 4 domains solved sequentially gives %.2f seconds with an accuracy of %.2e ' %  ( timeit.timeit('comb_pool4 = pool_4()', number=1,globals=globals()), diff_comb_pool4 ) )

    # plot the results
    time_spsolve = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.08, 0.13, 0.33, 0.52, 1.04, 2.05, 4.11, 8.35, 16.33, 32.55, 67.74]]
    time_comb2 = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.05, 0.11, 0.21, 0.4, 0.79, 1.60, 3.14, 6.25, 12.43, 25.93, 53.02]]
    time_comb4 = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.05, 0.09, 0.17, 0.33, 0.63, 1.21, 2.40, 4.97, 9.52, 20.72, 42.44]]

    plt.plot(time_spsolve[0], time_spsolve[1], label='current (sp_solve)', marker='+', color='gray')
    plt.plot(time_comb2[0], time_comb2[1], label='2 domains', marker='o')
    plt.plot(time_comb4[0], time_comb4[1], label='4 domains', marker='x', color='r')
    plt.xlabel('Time [s]')
    plt.ylabel('iterations')
    plt.legend()
    plt.show()
```
##  Performance 
![plot-decomposition](https://github.com/openearth/aeolis-python/assets/8195424/812fa3d7-2a13-46c5-b88d-01b899c4820d)

**Comment by manuGil** [comment]
# Timing Pool Sharing/Reuse
This example applies data decomposition (4 data parts), but a single pool is created and reused for all iterations. 

```python

def solve_matrix4(A,Ct_i):
    # solve the matrix
    result  = scipy.sparse.linalg.spsolve(A,Ct_i)

    return result

if __name__ == '__main__':
    from multiprocessing import Pool, Value, Array
    import time
    import matplotlib.pyplot as plt

    ######################################
    ## 4 domains with Pool Reuse
    ######################################                                                                                                  
     # Load data
    A = scipy.sparse.load_npz('parallell_solve/A.npz')
    yCt_i = np.load('parallell_solve/yCt_i.npy')

    # create a pool of workers
    pool = Pool()

    start_time = time.time()
    # using a look to simulate a loop over time steps
    for i in range(1):
        # devide matrices in 4 domains
        decs =    [ 
            (A[0:13545,0:13545], yCt_i.flatten()[0:13545]), # first dec
            (A[12040:25585,12040:25585], yCt_i.flatten()[12040:25585]), # second dec
            (A[24080:37625,24080:37625], yCt_i.flatten()[24080:37625]), # third dec
            (A[36120:45451,36120:45451], yCt_i.flatten()[36120:45451]) # fourth dec
        ]
        
        # using same pool 
        results = pool.starmap(solve_matrix4, decs)

        reshaped_results = [result.reshape(45,301) for result in results[:-1]]
        last_result = results[-1].reshape(31,301)

        # combine the results
        comb_pool4 = np.vstack((
            reshaped_results[0][0:41,:],reshaped_results[1][1:-5,:],reshaped_results[2][:-5,:],last_result[:,:]
            ))

    # close the pool and wait for the work to finish    
    pool.close()
    pool.join()

    end_time = time.time()

    print('decompostion using Pool with 4 domains solved sequentially gives %.2f seconds ' %  ( end_time-start_time ) )

    # plot the results
    time_spsolve = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.08, 0.13, 0.33, 0.52, 1.04, 2.05, 4.11, 8.35, 16.33, 32.55, 67.74]]
    time_comb2 = [
        [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024],
        [0.5, 0.11, 0.21, 0.4, 0.79, 1.60, 3.14, 6.25, 12.43, 25.93, 53.02]]
    time_comb4 = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.5, 0.09, 0.17, 0.33, 0.63, 1.21, 2.40, 4.97, 9.52, 20.72, 42.44]]
    time_comb4_pool = [[1,2,4, 8, 16, 32, 64, 128, 256, 512, 1024],[0.3, 0.06, 0.11, 0.22, 0.41, 0.86, 1.68, 3.45, 6.88, 13.70, 27.55]]

    plt.plot(time_spsolve[0], time_spsolve[1], label='current (sp_solve)', marker='+', color='gray')
    plt.plot(time_comb2[0], time_comb2[1], label='2 domains', marker='o')
    plt.plot(time_comb4[0], time_comb4[1], label='4 domains', marker='x', color='r')
    plt.plot(time_comb4_pool[0], time_comb4_pool[1], label='4 domains + reuse pool', marker='*', color='g')
    plt.xlabel('iterations')
    plt.ylabel('Time [s]')
    plt.legend()
    plt.show()
```

## Performance 
![plot-shared-pool](https://github.com/openearth/aeolis-python/assets/8195424/1dedc58e-47f0-4d92-93bb-d3dd6b673659)

**Comment by Sierd** [comment]
Thanks @manuGil !

I was wondering, could we open the pool when we initialize the model here:
https://github.com/openearth/aeolis-python/blob/fc75b06e9d0183e1e39f40b863d5a23a0252f7b8/aeolis/model.py#L2512

and then call/use it in solve_pieter here:
https://github.com/openearth/aeolis-python/blob/fc75b06e9d0183e1e39f40b863d5a23a0252f7b8/aeolis/model.py#L2164

and then close it when the model is finished running here:
https://github.com/openearth/aeolis-python/blob/fc75b06e9d0183e1e39f40b863d5a23a0252f7b8/aeolis/model.py#L2531

**Comment by Sierd** [comment]
@niketagrawal and @manuGil I made some adaptation to setup a parallel computations.
I needed to setup the pool outside the loop and pass it as an object to the function that uses the pool. This overcomes the need for setting up the pool inside the loop each timestap. It seems to work and the performance increases significantly (i am guessing 40%). 

@niketagrawal could you profile the code again using Barts Zandmotor simulation and give some feedback on this work?

I pushed the code in this branch.
https://github.com/openearth/aeolis-python/tree/parallell_update

**Comment by niketagrawal** [comment]
@Sierd cProfile reveals that the parallel version is approximately 36% faster than the current sequential version of the code when profiled on sandmotor example, however, the parallel version fails the test. 

The test `check netCDF file content` fails in the parallel version of the code.

```
========================================================================== short test summary info ==========================================================================
FAILED aeolis/tests/integration_tests/test_netCDF_file_content.py::test_netCDF_file_content - AssertionError: Array values of the parameter 'time' are expected to remain consistent across simulations for the same model parameter file for 2D Barchan dune
```

**Comment by niketagrawal** [comment]
- The speed up provided by the parallel version of the solver becomes significant when the size of the grids is large. The parallel version of the solver provides a speed of a factor of 2.7X for the sandmotor example with large grids. This is approximately twice the speed up achieved when tested on the sandmotor example with smaller grid size.

- The parallel version of the solver cannot be verified for accuracy of the results produced at the moment because it fails the test case in which the output of the generated netCDF file is compared against the expected results. The existing test asserts the results produced by the parallel version against the results of the sequential version of the simulation for the same use case. This may not be the correct approach to test the parallel version because of the partition of the domain into multiple overlapping domains.

- On the basis of the results obtained so far, the parallel version of the solver proves to be useful for specific cases only, therefore, generalizing its implementation for all cases is not the right approach. 
 
- It is concluded to keep this Proof of Concept implementation of hard coded style parallel implementation of the solver for future reference in this repository. The code changes will live in a separate feature branch in the repository.

**Comment by niketagrawal** [comment]
Code changes for the parallel version of the solver are in the branch [parallell_update](https://github.com/openearth/aeolis-python/tree/parallell_update)

Documentation for the proof of concept is being added via the pull request #127

**Comment by niketagrawal** [comment]
Closed with merging pull request #134
